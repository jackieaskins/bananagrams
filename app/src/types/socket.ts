import { BoardLocation } from "./board";
import { Game } from "./game";
import { PlayerStatus } from "./player";

export enum ClientToServerEventName {
  CreateGame = "createGame",
  JoinGame = "joinGame",
  KickPlayer = "kickPlayer",
  LeaveGame = "leaveGame",
  SetStatus = "setStatus",
  Split = "split",
  Peel = "peel",
  Dump = "dump",
  MoveTileFromHandToBoard = "moveTileFromHandToBoard",
  MoveTileFromBoardToHand = "moveTileFromBoardToHand",
  MoveTileOnBoard = "moveTileOnBoard",
  ShuffleHand = "shuffleHand",
}

export enum ServerToClientEventName {
  Notification = "notification",
  GameInfo = "gameInfo",
}

export type SocketCallback<T> = (
  error: { message: string } | null,
  args: T | null,
) => void;

export type SocketEventHandler<Event = null, CallbackArg = void> = (
  event: Event,
  callback?: SocketCallback<CallbackArg>,
) => void;

export type ClientToServerEvents = {
  [ClientToServerEventName.CreateGame]: SocketEventHandler<
    { gameName: string; username: string; isShortenedGame: boolean },
    Game
  >;
  [ClientToServerEventName.JoinGame]: SocketEventHandler<
    { gameId: string; username: string; isSpectator: boolean },
    Game
  >;
  [ClientToServerEventName.LeaveGame]: SocketEventHandler;

  [ClientToServerEventName.KickPlayer]: SocketEventHandler<{ userId: string }>;
  [ClientToServerEventName.SetStatus]: SocketEventHandler<{
    status: PlayerStatus;
  }>;

  [ClientToServerEventName.Split]: SocketEventHandler;
  [ClientToServerEventName.Peel]: SocketEventHandler;
  [ClientToServerEventName.Dump]: SocketEventHandler<{
    tileId: string;
    boardLocation: BoardLocation | null;
  }>;

  [ClientToServerEventName.ShuffleHand]: SocketEventHandler;

  [ClientToServerEventName.MoveTileFromHandToBoard]: SocketEventHandler<{
    tileId: string;
    boardLocation: BoardLocation;
  }>;
  [ClientToServerEventName.MoveTileFromBoardToHand]: SocketEventHandler<{
    boardLocation: BoardLocation;
  }>;
  [ClientToServerEventName.MoveTileOnBoard]: SocketEventHandler<{
    fromLocation: BoardLocation;
    toLocation: BoardLocation;
  }>;
};

export type ServerToClientEvents = {
  [ServerToClientEventName.Notification]: SocketEventHandler<{
    message: string;
  }>;
  [ServerToClientEventName.GameInfo]: SocketEventHandler<Game>;
};
