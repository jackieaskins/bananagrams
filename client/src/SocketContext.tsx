import React, { createContext, useContext, useEffect, useState } from 'react';
import io from 'socket.io-client';

export type Callback<T> = (error: Error | null, data: T) => void;

export type GameId = string;
export type GameName = string;
export type Username = string;

type SocketState = {
  currentUsername: Username;
  createGame: (
    { gameName, username }: { gameName: GameName; username: Username },
    callback?: Callback<GameId>
  ) => void;
  joinGame: (
    { gameId, username }: { gameId: GameId; username: Username },
    callback?: Callback<null>
  ) => void;
  leaveGame: (
    { gameId }: { gameId: GameId },
    callback?: Callback<null>
  ) => void;
};

const SOCKET_URL = 'http://localhost:5000';
let socket: SocketIOClient.Socket;

export const SocketContext = createContext<SocketState>({
  currentUsername: '',
  createGame: () => undefined,
  joinGame: () => undefined,
  leaveGame: () => undefined,
});

export const SocketProvider: React.FC<{}> = ({ children }) => {
  const [currentUsername, setCurrentUsername] = useState('');

  useEffect(() => {
    socket = io(SOCKET_URL);

    return (): void => {
      socket.emit('disconnect');
    };
  }, []);

  const createGame = (
    { gameName, username }: { gameName: GameName; username: Username },
    callback?: Callback<GameId>
  ): void => {
    const createGameCallback: Callback<GameId> = (error, gameId) => {
      if (!error) {
        setCurrentUsername(username);
      }
      callback?.(error, gameId);
    };
    socket.emit('createGame', { gameName, username }, createGameCallback);
  };

  const joinGame = (
    { gameId, username }: { gameId: GameId; username: Username },
    callback?: Callback<null>
  ): void => {
    const joinGameCallback: Callback<null> = (error) => {
      if (!error) {
        setCurrentUsername(username);
      }
      callback?.(error, null);
    };
    socket.emit('joinGame', { gameId, username }, joinGameCallback);
  };

  const leaveGame = (
    { gameId }: { gameId: GameId },
    callback?: Callback<null>
  ): void => {
    const leaveGameCallback: Callback<null> = (error) => {
      if (!error) {
        setCurrentUsername('');
      }
      callback?.(error, null);
    };
    socket.emit('leaveGame', { gameId }, leaveGameCallback);
  };

  return (
    <SocketContext.Provider
      value={{ currentUsername, createGame, joinGame, leaveGame }}
    >
      {children}
    </SocketContext.Provider>
  );
};

export const useSocket = (): SocketState => useContext(SocketContext);
