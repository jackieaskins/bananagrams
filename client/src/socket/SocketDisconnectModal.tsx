import { Button, Modal } from 'antd';
import { useCallback, useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { addDisconnectListener } from '.';

const SocketDisconnectModal = (): JSX.Element => {
  const [shouldShowModal, setShouldShowModal] = useState(false);
  const { push } = useHistory();

  useEffect(() => {
    addDisconnectListener(() => {
      setShouldShowModal(true);
    });
  }, []);

  const hideModal = useCallback(() => {
    push('/');
    setShouldShowModal(false);
  }, [push]);

  return (
    <Modal
      closable={false}
      footer={<Button onClick={hideModal}>Return home</Button>}
      title="Disconnected"
      visible={shouldShowModal}
    >
      You have been disconnected from the server
    </Modal>
  );
};

export default SocketDisconnectModal;
