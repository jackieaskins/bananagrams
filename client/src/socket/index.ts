import io from 'socket.io-client';

const SOCKET_URL = 'http://localhost:5000';

export default process.env.NODE_ENV === 'development' ? io(SOCKET_URL) : io();
