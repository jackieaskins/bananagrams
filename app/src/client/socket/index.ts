import { Socket, io } from "socket.io-client";
import { ClientToServerEvents, ServerToClientEvents } from "@/types/socket";

export const socket: Socket<ServerToClientEvents, ClientToServerEvents> = io({
  autoConnect: false,
});
