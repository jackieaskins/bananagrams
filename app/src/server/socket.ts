import { Server } from "socket.io";
import GameController from "./controllers/GameController";
import {
  ClientToServerEventName,
  ClientToServerEvents,
  ServerToClientEvents,
  SocketCallback,
} from "@/types/socket";

export function handler<E, T>(fn: (event: E) => T) {
  return function (event: E, callback?: SocketCallback<T>): void {
    try {
      const returnValue = fn(event) || null;
      callback?.(null, returnValue);
    } catch (e) {
      const { message } = e as Error;
      callback?.({ message }, null);
    }
  };
}

export function configureSocket(
  io: Server<ClientToServerEvents, ServerToClientEvents>,
): void {
  io.on("connection", (socket) => {
    const { id: userId } = socket;

    console.log(`New connection: ${userId}`);

    let gameController: GameController | undefined;

    function withController<E, T>(
      fn: (event: E, controller: GameController) => T,
    ) {
      return handler((event: E) => {
        if (!gameController) {
          throw new Error("Not in a game");
        }

        return fn(event, gameController);
      });
    }

    socket.on(
      ClientToServerEventName.CreateGame,
      handler(({ gameName, username, isShortenedGame }) => {
        gameController = GameController.createGame(
          gameName,
          username,
          isShortenedGame,
          io,
          socket,
        );
        return gameController.getCurrentGame().toJSON();
      }),
    );

    socket.on(
      ClientToServerEventName.JoinGame,
      handler(({ gameId, username, isSpectator }) => {
        gameController = GameController.joinGame(
          gameId,
          username,
          isSpectator,
          io,
          socket,
        );
        return gameController.getCurrentGame().toJSON();
      }),
    );

    socket.on(
      ClientToServerEventName.KickPlayer,
      withController(({ userId }, controller) => controller.kickPlayer(userId)),
    );

    socket.on(
      ClientToServerEventName.LeaveGame,
      withController((_event, controller) => {
        controller.leaveGame();
        gameController = undefined;
      }),
    );

    socket.on(
      ClientToServerEventName.SetStatus,
      withController(({ status }, controller) => controller.setStatus(status)),
    );

    socket.on(
      ClientToServerEventName.Split,
      withController((_event, controller) => controller.split()),
    );

    socket.on(
      ClientToServerEventName.Peel,
      withController((_event, controller) => controller.peel()),
    );

    socket.on(
      ClientToServerEventName.Dump,
      withController(({ tileId, boardLocation }, controller) =>
        controller.dump(tileId, boardLocation),
      ),
    );

    socket.on(
      ClientToServerEventName.MoveTilesFromHandToBoard,
      withController(({ tiles }, controller) =>
        controller.moveTilesFromHandToBoard(tiles),
      ),
    );

    socket.on(
      ClientToServerEventName.MoveTilesFromBoardToHand,
      withController(({ boardLocations }, controller) =>
        controller.moveTilesFromBoardToHand(boardLocations),
      ),
    );

    socket.on(
      ClientToServerEventName.MoveTilesOnBoard,
      withController(({ locations }, controller) =>
        controller.moveTilesOnBoard(locations),
      ),
    );

    socket.on(
      ClientToServerEventName.ShuffleHand,
      withController((_event, controller) => controller.shuffleHand()),
    );

    socket.on("disconnect", () => {
      gameController?.leaveGame();
      gameController = undefined;
      console.log(`${userId} disconnected`);
    });
  });
}
