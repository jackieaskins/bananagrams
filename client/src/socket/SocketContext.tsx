import { FunctionComponent, createContext, useContext, useEffect } from 'react';
import { Socket } from 'socket.io-client';

import socket from './index';

export type Callback<T> = (error: Error | null, data: T) => void;

type SocketState = {
  socket: Socket;
};

export const SocketContext = createContext<SocketState>({
  socket,
});

export const SocketProvider: FunctionComponent = ({ children }) => {
  useEffect(
    () => () => {
      socket.disconnect();
    },
    []
  );

  return (
    <SocketContext.Provider value={{ socket }}>
      {children}
    </SocketContext.Provider>
  );
};

export const useSocket = (): SocketState => useContext(SocketContext);
