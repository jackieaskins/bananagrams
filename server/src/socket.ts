import { Server } from 'socket.io';
import GameController from './controllers/GameController';

export const handler = (
  fn: () => Record<string, any> | void,
  callback?: (
    error: { message: string } | null,
    data: Record<string, any> | null
  ) => void
): void => {
  try {
    const returnValue = fn() || null;
    callback?.(null, returnValue);
  } catch ({ message }) {
    callback?.({ message }, null);
  }
};

export const configureSocket = (io: Server): void => {
  io.on('connection', (socket) => {
    const { id: userId } = socket;

    console.log(`New connection: ${userId}`);

    let gameController: GameController | undefined;

    const validateGameControllerExists = (): void => {
      if (!gameController) {
        throw new Error('Not in a game');
      }
    };

    socket.on(
      'createGame',
      ({ gameName, username, isShortenedGame }, callback) => {
        handler(() => {
          gameController = GameController.createGame(
            gameName,
            username,
            isShortenedGame,
            io,
            socket
          );
          return gameController.getCurrentGame().toJSON();
        }, callback);
      }
    );

    socket.on('joinGame', ({ gameId, username }, callback) => {
      handler(() => {
        gameController = GameController.joinGame(gameId, username, io, socket);
        return gameController.getCurrentGame().toJSON();
      }, callback);
    });

    socket.on('kickPlayer', ({ userId }, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).kickPlayer(userId);
      }, callback);
    });

    socket.on('leaveGame', ({}, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).leaveGame();
        gameController = undefined;
      }, callback);
    });

    socket.on('ready', ({ isReady }, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).setReady(isReady);
      }, callback);
    });

    socket.on('peel', ({}, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).peel();
      }, callback);
    });

    socket.on('dump', ({ tileId, boardLocation }, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).dump(tileId, boardLocation);
      }, callback);
    });

    socket.on(
      'moveTileFromHandToBoard',
      ({ tileId, boardLocation }, callback) => {
        handler(() => {
          validateGameControllerExists();
          (gameController as GameController).moveTileFromHandToBoard(
            tileId,
            boardLocation
          );
        }, callback);
      }
    );

    socket.on('moveTileFromBoardToHand', ({ boardLocation }, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).moveTileFromBoardToHand(
          boardLocation
        );
      }, callback);
    });

    socket.on('moveAllTilesFromBoardToHand', ({}, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).moveAllTilesFromBoardToHand();
      }, callback);
    });

    socket.on('moveTileOnBoard', ({ fromLocation, toLocation }, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).moveTileOnBoard(
          fromLocation,
          toLocation
        );
      }, callback);
    });

    socket.on('shuffleHand', ({}, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).shuffleHand();
      }, callback);
    });

    socket.on('disconnect', () => {
      gameController?.leaveGame();
      gameController = undefined;
      console.log(`${userId} disconnected`);
    });
  });
};
