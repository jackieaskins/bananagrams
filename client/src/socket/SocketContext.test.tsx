/* eslint-disable @typescript-eslint/ban-ts-comment */
import { useEffect, useContext } from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { SocketProvider, useSocket, SocketContext } from "./SocketContext";
import socket from ".";

jest.mock("react", () => ({
  ...(jest.requireActual("react") as Record<string, unknown>),
  useContext: jest.fn(),
  useEffect: jest.fn().mockImplementation((f) => f()),
}));
jest.mock("./index", () => ({
  on: jest.fn(),
  emit: jest.fn(),
}));
const mockEnqueueSnackbar = jest.fn();
jest.mock("notistack", () => ({
  useSnackbar: (): any => ({ enqueueSnackbar: mockEnqueueSnackbar }),
}));

describe("SocketContext", () => {
  describe("<SocketProvider />", () => {
    let component: ShallowWrapper;

    beforeEach(() => {
      component = shallow(<SocketProvider>children</SocketProvider>);
    });

    test("renders provider properly", () => {
      expect(component).toMatchSnapshot();
    });

    test("listens on socket for notifications", () => {
      expect(socket.on).toHaveBeenCalledWith(
        "notification",
        expect.any(Function),
      );
    });

    test("enqueues message to snackbar on message received", () => {
      const message = "Message";

      // @ts-ignore
      socket.on.mock.calls[0][1]({ message });

      expect(mockEnqueueSnackbar).toHaveBeenCalledWith(message);
    });

    test("disconnects from socket on dismount", () => {
      // @ts-ignore
      useEffect.mock.calls[0][0]()();

      expect(socket.emit).toHaveBeenCalledWith("disconnect");
    });
  });

  describe("useSocket", () => {
    test("calls useContext with SocketContext", () => {
      useSocket();
      expect(useContext).toHaveBeenCalledWith(SocketContext);
    });
  });
});
