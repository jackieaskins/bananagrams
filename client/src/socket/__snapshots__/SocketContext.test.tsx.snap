// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocketContext <SocketProvider /> renders provider properly 1`] = `
<ContextProvider
  value={
    Object {
      "socket": Object {
        "disconnect": [MockFunction],
        "emit": [MockFunction],
        "on": [MockFunction],
      },
    }
  }
>
  children
</ContextProvider>
`;
